scpi-pps: Add Siglent SPD3303 series PSU support

tjko opened the pull request:
This adds support to Siglent SPD3303C / SPD3303X / SPD3303X-E programmable DC power supplies.
Functionality was tested both via LAN (tcp-raw) and USB (USBTMC) using sigrok-cli and SmuView.

Siglent SCPI implementation is "problematic" so I implemented "quirks" in SCPI framework to allow supporting
instruments (vendors) that can't seem to follow SCPI/IEEE488.2 specifications...
This way SCPI behavior doesn't need to change for devices that comply with SCPI, but framework still can be used
with devices that require "tweaks": 9cf9a57acfcdb6230a8d176da10f7f80a95cca3b

(replaces https://github.com/sigrokproject/libsigrok/pull/75)

tjko commented:
Wiki entry created: https://sigrok.org/wiki/Siglent_SPD3303_series

LeSpocky commented:
> Wiki entry created: https://sigrok.org/wiki/Siglent_SPD3303_series

See https://sigrok.org/wiki/Siglent_SPD3303C/Info for `lsusb` output.

LeSpocky commented:
I did a quick test on Debian GNU/Linux 10 (buster) with the SPD3303C. Output of --driver --show is this:

```
adahl@ada ~/sr % LD_LIBRARY_PATH=${HOME}/sr/lib ./bin/sigrok-cli --driver=scpi-pps --channel-group 1 --config "enabled=on" --show      
Driver functions:
    Power supply
Scan options:
    conn
    serialcomm
scpi-pps - Siglent Technologies SPD3303C 1.02.01.01.03R2 [S/N: SPD3EDCD3R1506] with 6 channels: V1 I1 P1 V2 I2 P2
Channel groups:
    1: channels V1 I1 P1
    2: channels V2 I2 P2
Supported configuration options on channel group 1:
    regulation: CV (current)
    voltage: 23.990000 (current)
    voltage_target: 0.000000, 32.000000, 0.010000
    current: 0.050000 (current)
    current_limit: 0.000000, 3.200000, 0.010000
    enabled: on (current), off
```

This worked fine:

```
adahl@ada ~/sr % LD_LIBRARY_PATH=${HOME}/sr/lib ./bin/sigrok-cli --driver=scpi-pps --channel-group 1 --config "enabled=on" --set 
```

Most things I tried, ended up in the instrument beeping and the console showing an error message repeatedly. See below:

```
adahl@ada ~/sr % LD_LIBRARY_PATH=${HOME}/sr/lib ./bin/sigrok-cli --scan
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
^C
```

beep, beep, beep, …

---
```
adahl@ada ~/sr % LD_LIBRARY_PATH=${HOME}/sr/lib ./bin/sigrok-cli --driver=scpi-pps --time 3s
V1: 0 mV DC
I1: 0 mA DC
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
^C
```

beep, beep, beep, …

---
```
adahl@ada ~/sr % LD_LIBRARY_PATH=${HOME}/sr/lib ./bin/sigrok-cli --driver=scpi-pps --channel-group 1 --config "enabled=on" --continuous
V1: 23.990 V DC
I1: 50 mA DC
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
```

beep, beep, beep, …

tjko commented:
Looks like this model has same (or similar) buggy firmware with broken USBTMC implementation (as SPD3033X-E).
If you just do "--scan" and not use "-d scpi-pps" it seems unit can stop responding properly. Does it recover if unplug and use only commands with "-d scpi-pps"?  

With SPD3033X-E unit worked much more reliably via LAN, but I guess this model does not have LAN port?

LeSpocky commented:
> Looks like this model has same (or similar) buggy firmware with broken USBTMC implementation (as SPD3033X-E).

Don't know, did not look into it or even try firmware update. I would try to avoid such an upgrade, my boss would not be happy if anything goes wrong with that.

> If you just do "--scan" and not use "-d scpi-pps" it seems unit can stop responding properly. Does it recover if unplug and use only commands with "-d scpi-pps"?

No, it does not recover.

> With SPD3033X-E unit worked much more reliably via LAN, but I guess this model does not have LAN port?

Correct. The SPD3303C has _no_ LAN port.

tjko commented:
Merge conflicts fixed in new branch: https://github.com/tjko/libsigrok/tree/siglent-scpi-v4

tjko commented:
Updated branch with some cleanup:
https://github.com/tjko/libsigrok/tree/siglent-scpi-v5

Also removed SPD3303C / SPD3303D support, as these units don't seem to work with current scpi_usbtmc support in Sigrok:

```
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
sr: scpi_usbtmc: USBTMC bulk in transfer error: LIBUSB_ERROR_TIMEOUT.
```

biot commented:
Sorry this never got reviewed, @tjko. Your `siglent-scpi-v5` branch looks good, only some minor style fixes. Could I ask you to rebase it on master, and push it here? I'd like to see this get merged.
